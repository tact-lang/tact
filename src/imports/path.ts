import { throwInternalCompilerError } from "../error/errors";
import { repeat } from "../utils/array";

// Witness tag. Do not use, do not export.
const pathTag = Symbol("path");

/**
 * Safe relative path
 */
export type RelativePath = {
    /**
     * Number of "../" in front of path
     */
    readonly stepsUp: number;
    /**
     * /-separated strings that go after optional ../
     */
    readonly segments: readonly string[];
    /**
     * Proof that path was created by RelativePath constructor
     */
    readonly [pathTag]: true;
};

/**
 * Constructor for relative paths
 */
const RelativePath = (
    stepsUp: number,
    segments: readonly string[],
): RelativePath => {
    if (stepsUp < 0) {
        throwInternalCompilerError("Negative number of ../ in path");
    }
    const result: RelativePath = {
        [pathTag]: true,
        stepsUp,
        segments: Object.freeze(segments),
    };
    return Object.freeze(result);
};

/**
 * Convert raw string with relative POSIX path into RelativePath
 */
export const fromString = (raw: string): RelativePath => {
    return raw.split("/").map(parseSegment).reduce(appendPath, emptyPath);
};

/**
 * Convert RelativePath to string.
 */
export const asString = ({ stepsUp, segments }: RelativePath): string => {
    return [...repeat("..", stepsUp), ...segments].join("/");
};

/**
 * Empty path, equivalent to "."
 */
export const emptyPath = RelativePath(0, []);

/**
 * Combine two relative paths
 */
const appendPath = (left: RelativePath, right: RelativePath): RelativePath => {
    const delta = right.stepsUp - left.segments.length;
    return RelativePath(left.stepsUp + Math.max(0, delta), [
        ...left.segments.slice(0, Math.max(0, -delta)),
        ...right.segments,
    ]);
};

const parseSegment = (segment: string): RelativePath => {
    if (segment === "..") {
        return RelativePath(1, []);
    }
    if (segment === "." || segment === "") {
        return emptyPath;
    }
    return RelativePath(0, [segment]);
};
