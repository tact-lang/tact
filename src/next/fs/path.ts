/**
 * Safe relative path
 */
export type RelativePath = {
    /**
     * Number of "../" in front of path
     */
    readonly stepsUp: number;
    /**
     * /-separated strings that go after optional ../
     */
    readonly segments: readonly string[];
};
