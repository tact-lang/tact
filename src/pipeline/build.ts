import { beginCell, Cell, Dictionary } from '@ton/core';
import { decompileAll } from '@tact-lang/opcode';
import { writeTypescript } from '../bindings/writeTypescript';
import { featureEnable } from '../config/features';
import { ConfigProject } from "../config/parseConfig";
import { CompilerContext } from "../context";
import { funcCompile } from '../func/funcCompile';
import { writeReport } from '../generator/writeReport';
import { getRawAST } from '../grammar/store';
import files from '../imports/stdlib';
import { consoleLogger, TactLogger } from '../logger';
import { PackageFileFormat } from '../packaging/fileFormat';
import { packageCode } from '../packaging/packageCode';
import { createABITypeRefFromTypeRef } from '../types/resolveABITypeRef';
import { getContracts, getType } from '../types/resolveDescriptors';
import { errorToString } from '../utils/errorToString';
import { posixNormalize } from '../utils/filePath';
import { createVirtualFileSystem } from '../vfs/createVirtualFileSystem';
import { VirtualFileSystem } from '../vfs/VirtualFileSystem';
import { compile } from './compile';
import { precompile } from "./precompile";
import { getCompilerVersion } from './version';

export async function build(args: {
    config: ConfigProject,
    project: VirtualFileSystem,
    stdlib: string | VirtualFileSystem,
    logger?: TactLogger | null | undefined
}) {

    const { config, project } = args;
    const stdlib = (typeof args.stdlib === 'string') ? createVirtualFileSystem(args.stdlib, files) : args.stdlib;
    const logger: TactLogger = args.logger || consoleLogger;

    // Configure context
    let ctx: CompilerContext = new CompilerContext({ shared: {} });
    const cfg: string = JSON.stringify({
        entrypoint: config.path,
        options: (config.options || {})
    });
    if (config.options) {
        if (config.options.debug) {
            logger.error('   > ðŸ‘€ Enabling debug');
            ctx = featureEnable(ctx, 'debug');
        }
        if (config.options.masterchain) {
            logger.error('   > ðŸ‘€ Enabling masterchain');
            ctx = featureEnable(ctx, 'masterchain');
        }
        if (config.options.external) {
            logger.error('   > ðŸ‘€ Enabling external');
            ctx = featureEnable(ctx, 'external');
        }
        if (config.options.experimental && config.options.experimental.inline) {
            logger.error('   > ðŸ‘€ Enabling inline');
            ctx = featureEnable(ctx, 'inline');
        }
    }

    // Precompile
    try {
        ctx = precompile(ctx, project, stdlib, config.path);
    } catch (e) {
        logger.error('Tact compilation failed');
        logger.error(errorToString(e));
        return false;
    }

    // Compile contracts
    let ok = true;
    const built: {
        [key: string]: {
            codeBoc: Buffer,
            // codeFunc: string,
            // codeFift: string,
            // codeFiftDecompiled: string,
            abi: string
        }
    } = {};
    for (const contract of getContracts(ctx)) {
        const pathAbi = project.resolve(config.output, config.name + '_' + contract + ".abi");


        const pathCodeBoc = project.resolve(config.output, config.name + '_' + contract + ".code.boc");
        const pathCodeFif = project.resolve(config.output, config.name + '_' + contract + ".code.fif");
        const pathCodeFifDec = project.resolve(config.output, config.name + '_' + contract + ".code.rev.fif");
        let codeFc: { path: string, content: string }[];
        let codeEntrypoint: string;

        // Compiling contract to func
        logger.log('   > ' + contract + ': tact compiler');
        let abi: string;
        try {
            const res = await compile(ctx, contract, config.name + '_' + contract);
            for (const files of res.output.files) {
                const ffc = project.resolve(config.output, files.name);
                project.writeFile(ffc, files.code);
            }
            project.writeFile(pathAbi, res.output.abi);
            abi = res.output.abi;
            codeFc = res.output.files.map((v) => ({ path: posixNormalize(project.resolve(config.output, v.name)), content: v.code }));
            codeEntrypoint = res.output.entrypoint;
        } catch (e) {
            logger.error('Tact compilation failed');
            logger.error(errorToString(e));
            ok = false;
            continue;
        }

        // Compiling contract to TVM
        logger.log('   > ' + contract + ': func compiler');
        let codeBoc: Buffer;
        try {
            const stdlibPath = stdlib.resolve('stdlib.fc');
            const stdlibCode = stdlib.readFile(stdlibPath).toString();
            const stdlibExPath = stdlib.resolve('stdlib_ex.fc');
            const stdlibExCode = stdlib.readFile(stdlibExPath).toString();
            const c = await funcCompile({
                entries: [
                    stdlibPath,
                    stdlibExPath,
                    posixNormalize(project.resolve(config.output, codeEntrypoint))
                ],
                sources: [{
                    path: stdlibPath,
                    content: stdlibCode
                }, {
                    path: stdlibExPath,
                    content: stdlibExCode,
                },
                ...codeFc
                ],
                logger
            });
            if (!c.ok) {
                logger.error(c.log);
                ok = false;
                continue;
            }
            project.writeFile(pathCodeFif, c.fift);
            project.writeFile(pathCodeBoc, c.output);
            codeBoc = c.output;
        } catch (e) {
            logger.error('FunC compiler crashed');
            logger.error(errorToString(e));
            ok = false;
            continue;
        }

        // Fift decompiler for generated code debug
        logger.log('   > ' + contract + ': fift decompiler');
        let codeFiftDecompiled: string;
        try {
            codeFiftDecompiled = decompileAll({ src: codeBoc });
            project.writeFile(pathCodeFifDec, codeFiftDecompiled);
        } catch (e) {
            logger.error('Fift decompiler crashed');
            logger.error(errorToString(e));
            ok = false;
            continue;
        }

        // Add to built map
        built[contract] = {
            // codeFunc,
            codeBoc,
            // codeFift,
            // codeFiftDecompiled,
            abi
        };
    }
    if (!ok) {
        logger.log('ðŸ’¥ Compilation failed. Skipping packaging');
        return false;
    }

    // Package
    logger.log('   > Packaging');
    const contracts = getContracts(ctx);
    const packages: PackageFileFormat[] = [];
    for (const contract of contracts) {
        logger.log('   > ' + contract);
        const artifacts = built[contract];
        if (!artifacts) {
            logger.error('   > ' + contract + ': no artifacts found');
            return false;
        }

        // System cell
        const depends = Dictionary.empty(Dictionary.Keys.Uint(16), Dictionary.Values.Cell());
        const ct = getType(ctx, contract);
        depends.set(ct.uid, Cell.fromBoc(built[ct.name].codeBoc)[0]); // Mine
        for (const c of ct.dependsOn) {
            const cd = built[c.name];
            if (!cd) {
                logger.error('   > ' + cd + ': no artifacts found');
                return false;
            }
            depends.set(c.uid, Cell.fromBoc(cd.codeBoc)[0]);
        }
        const systemCell = beginCell().storeDict(depends).endCell();

        // Collect sources
        const sources: { [key: string]: string } = {};
        const rawAst = getRawAST(ctx);
        for (const source of [...rawAst.funcSources, ...rawAst.sources]) {
            if (source.path.startsWith(project.root) && !source.path.startsWith(stdlib.root)) {
                sources[source.path.slice(project.root.length)] = Buffer.from(source.code).toString('base64');
            }
        }

        // Package
        const pkg: PackageFileFormat = {
            name: contract,
            abi: artifacts.abi,
            code: artifacts.codeBoc.toString('base64'),
            init: {
                kind: 'direct',
                args: getType(ctx, contract).init!.args.map((v) => ({ name: v.name, type: createABITypeRefFromTypeRef(v.type, v.ref) })),
                prefix: {
                    bits: 1,
                    value: 0,
                },
                deployment: {
                    kind: 'system-cell',
                    system: systemCell.toBoc().toString('base64')
                },
            },
            sources,
            compiler: {
                name: 'tact',
                version: getCompilerVersion(),
                parameters: cfg
            }
        };
        const pkgData = packageCode(pkg);
        const pathPkg = project.resolve(config.output, config.name + '_' + contract + ".pkg");
        project.writeFile(pathPkg, pkgData);
        packages.push(pkg);
    }

    // Bindings
    logger.log('   > Bindings');
    for (const pkg of packages) {
        logger.log('   > ' + pkg.name);
        if (pkg.init.deployment.kind !== 'system-cell') {
            logger.error('   > ' + pkg.name + ': unsupported deployment kind ' + pkg.init.deployment.kind);
            return false;
        }
        try {
            const bindingsServer = writeTypescript(JSON.parse(pkg.abi), {
                code: pkg.code,
                prefix: pkg.init.prefix,
                system: pkg.init.deployment.system,
                args: pkg.init.args
            });
            project.writeFile(project.resolve(config.output, config.name + '_' + pkg.name + ".ts"), bindingsServer);
        } catch (e) {
            logger.error('Bindings compiler crashed');
            logger.error(errorToString(e));
            return false;
        }
    }

    // Reports
    logger.log('   > Reports');
    for (const pkg of packages) {
        logger.log('   > ' + pkg.name);
        try {
            const report = writeReport(ctx, pkg);
            const pathBindings = project.resolve(config.output, config.name + '_' + pkg.name + ".md");
            project.writeFile(pathBindings, report);
        } catch (e) {
            logger.error('Report generation crashed');
            logger.error(errorToString(e));
            return false;
        }
    }

    return true;
}