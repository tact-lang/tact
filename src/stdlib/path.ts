import path from "path";
import { posixNormalize } from "../utils/filePath";

export const stdlibPath = posixNormalize(path.join(__dirname, "stdlib"));
