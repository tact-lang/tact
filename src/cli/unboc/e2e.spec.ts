import { join, normalize } from "path";
import { makeCodegen, runCommand } from "../test-util.build";

// disable tests on windows
const testWin = process.platform !== "win32" ? test : test.skip;

const tact = (args: string) => {
    const tactPath = normalize(
        join(__dirname, "..", "..", "..", "bin", "tact.js"),
    );
    const command = `node ${tactPath} ${args}`;
    return runCommand(command);
};
const unboc = (args: string) => {
    const unbocPath = normalize(
        join(__dirname, "..", "..", "..", "bin", "unboc.js"),
    );
    const command = `node ${unbocPath} ${args}`;
    return runCommand(command);
};

const codegen = makeCodegen(join(__dirname, "output"));

const goodContract = `
contract Test {
    get fun greeting(): String {
        return "hello world";
    }
}
`;

describe("unboc foo.boc", () => {
    testWin("Exits with correct code", async () => {
        const r = await codegen.config(`unboc`, goodContract, {});
        await tact(`-c ${r.config}`);
        const result = await unboc(r.outputPath("code.boc"));
        expect(result).toMatchObject({ kind: "exited", code: 0 });
    });
});
