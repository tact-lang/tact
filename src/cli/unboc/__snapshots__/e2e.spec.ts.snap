// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unboc foo.boc Default run 1`] = `
{
  "code": 0,
  "kind": "exited",
  "stderr": "",
  "stdout": ""Asm.fif" include
PROGRAM{
  DECLPROC recv_internal
  78250 DECLMETHOD ?fun_78250
  DECLPROC ?fun_ref_92183b49329bb4e4
  recv_internal PROC:<{
    DROP
    DROP
    CTOS
    TWO
    SDSKIPFIRST
    1 LDI
    1 LDI
    LDMSGADDR
    OVER
    s3 s4 XCHG
    s5 s5 XCHG2
    4 TUPLE
    1 SETGLOB
    SWAP
    2 SETGLOB
    PUSHROOT
    CTOS
    1 LDI
    DROP
    <{
      NULL
    }> PUSHCONT
    <{
      NULL
    }> PUSHCONT
    IFELSE
    DROP
    IFRET
    130 THROW
  }>
  ?fun_78250 PROC:<{
    PUSHROOT
    CTOS
    1 LDI
    DROP
    <{
      NULL
    }> PUSHCONT
    <{
      NULL
    }> PUSHCONT
    IFELSE
    ?fun_ref_92183b49329bb4e4 INLINECALLDICT
    NIP
  }>
  ?fun_ref_92183b49329bb4e4 PROCREF:<{
    x{68656C6C6F20776F726C64} PUSHSLICE
  }>
}END>c
",
}
`;

exports[`unboc foo.boc With bitcode 1`] = `
{
  "code": 0,
  "kind": "exited",
  "stderr": "",
  "stdout": ""Asm.fif" include
PROGRAM{
  DECLPROC recv_internal
  78250 DECLMETHOD ?fun_78250
  DECLPROC ?fun_ref_92183b49329bb4e4
  recv_internal PROC:<{
    DROP                                          // 0x3 0
    DROP                                          // 0x3 0
    CTOS                                          // 0xD0
    TWO                                           // 0x7 2
    SDSKIPFIRST                                   // 0xD721
    1 LDI                                         // 0xD2 00
    1 LDI                                         // 0xD2 00
    LDMSGADDR                                     // 0xFA40
    OVER                                          // 0x2 1
    s3 s4 XCHG                                    // 0x10 3 4
    s5 s5 XCHG2                                   // 0x50 5 5
    4 TUPLE                                       // 0x6F0 4
    1 SETGLOB                                     // 0xF87_ 0C_
    SWAP                                          // 0x0 1
    2 SETGLOB                                     // 0xF87_ 14_
    PUSHROOT                                      // 0xED4 4
    CTOS                                          // 0xD0
    1 LDI                                         // 0xD2 00
    DROP                                          // 0x3 0
    <{
      NULL                                        // 0x6D
    }> PUSHCONT                                   // 0x9 6D
    <{
      NULL                                        // 0x6D
    }> PUSHCONT                                   // 0x9 6D
    IFELSE                                        // 0xE2
    DROP                                          // 0x3 0
    IFRET                                         // 0xDC
    130 THROW                                     // 0xF2C4_ 105_
  }>
  ?fun_78250 PROC:<{
    PUSHROOT                                      // 0xED4 4
    CTOS                                          // 0xD0
    1 LDI                                         // 0xD2 00
    DROP                                          // 0x3 0
    <{
      NULL                                        // 0x6D
    }> PUSHCONT                                   // 0x9 6D
    <{
      NULL                                        // 0x6D
    }> PUSHCONT                                   // 0x9 6D
    IFELSE                                        // 0xE2
    ?fun_ref_92183b49329bb4e4 INLINECALLDICT      // 0x
    NIP                                           // 0x3 1
  }>
  ?fun_ref_92183b49329bb4e4 PROCREF:<{
    x{68656C6C6F20776F726C64} PUSHSLICE           // 0x8B 68656C6C6F20776F726C64
  }>
}END>c
",
}
`;

exports[`unboc foo.boc Without aliases 1`] = `
{
  "code": 0,
  "kind": "exited",
  "stderr": "",
  "stdout": ""Asm.fif" include
PROGRAM{
  DECLPROC recv_internal
  78250 DECLMETHOD ?fun_78250
  DECLPROC ?fun_ref_92183b49329bb4e4
  recv_internal PROC:<{
    s0 POP
    s0 POP
    CTOS
    2 PUSHINT
    SDSKIPFIRST
    1 LDI
    1 LDI
    LDMSGADDR
    s1 PUSH
    s3 s4 XCHG
    s5 s5 XCHG2
    4 TUPLE
    1 SETGLOB
    s0 s1 XCHG
    2 SETGLOB
    c4 PUSHCTR
    CTOS
    1 LDI
    s0 POP
    <{
      NULL
    }> PUSHCONT
    <{
      NULL
    }> PUSHCONT
    IFELSE
    s0 POP
    IFRET
    130 THROW
  }>
  ?fun_78250 PROC:<{
    c4 PUSHCTR
    CTOS
    1 LDI
    s0 POP
    <{
      NULL
    }> PUSHCONT
    <{
      NULL
    }> PUSHCONT
    IFELSE
    ?fun_ref_92183b49329bb4e4 INLINECALLDICT
    s1 POP
  }>
  ?fun_ref_92183b49329bb4e4 PROCREF:<{
    x{68656C6C6F20776F726C64} PUSHSLICE
  }>
}END>c
",
}
`;

exports[`unboc foo.boc Without refs 1`] = `
{
  "code": 0,
  "kind": "exited",
  "stderr": "",
  "stdout": ""Asm.fif" include
PROGRAM{
  DECLPROC recv_internal
  78250 DECLMETHOD ?fun_78250
  recv_internal PROC:<{
    DROP
    DROP
    CTOS
    TWO
    SDSKIPFIRST
    1 LDI
    1 LDI
    LDMSGADDR
    OVER
    s3 s4 XCHG
    s5 s5 XCHG2
    4 TUPLE
    1 SETGLOB
    SWAP
    2 SETGLOB
    PUSHROOT
    CTOS
    1 LDI
    DROP
    <{
      NULL
    }> PUSHCONT
    <{
      NULL
    }> PUSHCONT
    IFELSE
    DROP
    IFRET
    130 THROW
  }>
  ?fun_78250 PROC:<{
    PUSHROOT
    CTOS
    1 LDI
    DROP
    <{
      NULL
    }> PUSHCONT
    <{
      NULL
    }> PUSHCONT
    IFELSE
    <{
      x{68656C6C6F20776F726C64} PUSHSLICE
    }>c CALLREF
    NIP
  }>
}END>c
",
}
`;
